src/ucs/datastruct/bitmap.h

```diff
@@ -54,7 +70,90 @@ typedef uint64_t ucs_bitmap_word_t;
     (-2ull << (uint64_t)((_index) & (UCS_BITMAP_BITS_IN_WORD - 1)))
 
 
-/*
+#define _UCS_BITMAP_FOR_EACH_WORD(_bitmap, _word_index) \
+    for (_word_index = 0; _word_index < UCS_BITMAP_WORDS(_bitmap); _word_index++)
+
+
+/**
+ * Perform inplace bitwise NOT of a bitmap
+ *
+ * @param _bitmap Negate this bitmap
+ */
+#define UCS_BITMAP_INPLACE_NOT(_bitmap) \
+    { \
+        size_t _word_index; \
+        _UCS_BITMAP_FOR_EACH_WORD(_bitmap, _word_index) { \
+            _UCS_BITMAP_WORD(_bitmap, _word_index) = ~_UCS_BITMAP_WORD(_bitmap, _word_index); \
+        } \
+    }
+
+
+#define _UCS_BITMAP_INPLACE_OP(_bitmap1, _bitmap2, _op) \
+    { \
+        size_t           _word_index; \
+        typeof(_bitmap1) _bitmap2_copy = _bitmap2; \
+        _UCS_BITMAP_FOR_EACH_WORD(_bitmap1, _word_index) { \
+            _UCS_BITMAP_WORD(_bitmap1, _word_index) = _UCS_BITMAP_WORD(_bitmap1, _word_index) _op \
+                                     _UCS_BITMAP_WORD(_bitmap2_copy, _word_index); \
+        } \
+    }
+
+
+/**
+ * Perform inplace bitwise AND of 2 bitmaps, storing the result in the first one
+ *
+ * @param _bitmap1 First operand
+ * @param _bitmap1 Second operand
+ */
+#define UCS_BITMAP_INPLACE_AND(_bitmap1, _bitmap2) \
+    _UCS_BITMAP_INPLACE_OP(_bitmap1, _bitmap2, &)
+
+
+/**
+ * Perform inplace bitwise OR of 2 bitmaps, storing the result in the first one
+ *
+ * @param _bitmap1 First operand
+ * @param _bitmap1 Second operand
+ */
+#define UCS_BITMAP_INPLACE_OR(_bitmap1, _bitmap2) \
+    _UCS_BITMAP_INPLACE_OP(_bitmap1, _bitmap2, |)
+
+
+/**
+ * Perform inplace bitwise XOR of 2 bitmaps, storing the result in the first one
+ *
+ * @param _bitmap1 First operand
+ * @param _bitmap1 Second operand
+ */
+#define UCS_BITMAP_INPLACE_XOR(_bitmap1, _bitmap2) \
+    _UCS_BITMAP_INPLACE_OP(_bitmap1, _bitmap2, ^)
+
+
+static inline bool _ucs_bitmap_is_zero(const void *bitmap, size_t words)
```

Comment:
1. UCS_F_ALWAYS_INLINE
2. words -> num_words
3. let's move this func definition after all the macros
4. maybe remove leading `_`

Reply:
Done 3/4. I don't think we shall allow using that func from outside.

Reply:
- move type instantiation to end